lightning/src/util/persist.rs

```diff
@@ -243,10 +254,873 @@ where
+	pub fn cleanup_stale_updates<ES: Deref + Clone, SP: Deref + Clone>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		search_depth: u64,
+	) -> io::Result<()>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		let monitor_list = self.kv_store.list(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+		)?;
+		for maybe_monitor_name in monitor_list.into_iter().map(MonitorName::new) {
+			let monitor_name = maybe_monitor_name?;
+			let (_, monitor) = self.read_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			let end = monitor.get_latest_update_id();
+			let start = end.saturating_sub(search_depth);
+			self.lazy_delete_updates(&monitor_name, start, end)?;
+		}
+		Ok(())
+	}
+}
+
+impl<ChannelSigner: WriteableEcdsaChannelSigner, K: Deref, L: Deref> Persist<ChannelSigner>
+	for MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	/// Persists a new channel. This means writing the entire monitor to the
+	/// provided [`KVStore`].
+	fn persist_new_channel(
+		&self,
+		funding_txo: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		_monitor_update_call_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		let monitor_name = match MonitorName::try_from(funding_txo) {
+			Ok(n) => n,
+			Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+		};
+		let key = monitor_name.as_str();
+		let mut monitor_bytes = Vec::with_capacity(
+			MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() + monitor.serialized_length(),
+		);
+		monitor_bytes.extend_from_slice(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL);
+		monitor.write(&mut monitor_bytes).unwrap();
+		match self.kv_store.write(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			key,
+			&monitor_bytes,
+		) {
+			Ok(_) => {
+				if let Some(old_update_id) = self
+					.last_monitor_update_id
+					.borrow_mut()
+					.insert(monitor_name, monitor.get_latest_update_id())
+				{
+					// Typically, we'll clean up between the last two known full monitors
+					let start = old_update_id;
+					let mut end = monitor.get_latest_update_id();
+					if monitor.get_latest_update_id() == CLOSED_CHANNEL_UPDATE_ID {
+						// We don't want to clean the rest of u64, so just do possible pending updates
+						end = old_update_id + self.maximum_pending_updates;
+					}
+					if let Err(e) =
+						self.lazy_delete_updates(&MonitorName::from(funding_txo), start, end)
+					{
+						log_error!(
+							self.logger,
+							"error cleaning up channel monitor updates {}",
+							e
+						);
+					};
+				};
+				chain::ChannelMonitorUpdateStatus::Completed
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"error writing channel monitor {}",
+					e
+				);
+				chain::ChannelMonitorUpdateStatus::PermanentFailure
```

Comment:
Done. #2562 is interesting though, treating all errors as transient means I can't express an actual permanent failure now, since there is no semantic for it?

Reply:
Done. #2562 is interesting though -- in another comment you mention this becoming fallible, is that part of the idea? The status enum will reflect either success or transient error, and the `Err` will be there for actually permanent failures?